sweyer_plugin/lib/sweyer_plugin.dart

```diff
@@ -0,0 +1,183 @@
+
+/// A playlist from the native media store.
+abstract class MediaStorePlaylist {
+  /// The id of this playlist in the native media store.
+  int get id;
+
+  /// A list of ids of songs that this playlist contains.
+  List<int> get songIds;
+}
+
+/// A genre from the native media store.
+abstract class MediaStoreGenre {}
+
+abstract class SweyerPluginException with EquatableMixin {
+  const SweyerPluginException();
+  String get value;
+
+  @override
+  List<Object?> get props => [value];
+
+  /// Generic error.
+  static const unexpected = SweyerMethodChannelException.unexpected;
+
+  /// On Android 30 requests like `MediaStore.createDeletionRequest` require
+  /// calling `startIntentSenderForResult`, which might throw this exception.
+  static const intentSender = SweyerMethodChannelException.intentSender;
+
+  static const io = SweyerMethodChannelException.io;
+
+  /// API is unavailable on current SDK level.
+  static const sdk = SweyerMethodChannelException.sdk;
+
+  /// Operation cannot be performed because there's no such playlist.
+  static const playlistNotExists = SweyerMethodChannelException.playlistNotExists;
+}
+
+const _uuid = Uuid();
+
+/// Class to cancel the [ContentChannel.loadAlbumArt].
+class CancellationSignal {
+  CancellationSignal() : _id = _uuid.v4();
+  final String _id;
+
+  /// Cancel loading of an album art.
+  Future<void> cancel() => SweyerPluginPlatform.instance.cancelAlbumArtLoad(id: _id);
+}
+
+abstract class SweyerPlugin {
+  static Future<Uint8List?> loadAlbumArt({
+    required String uri,
+    required Size size,
+    required CancellationSignal signal,
+  }) async =>
+      await SweyerPluginPlatform.instance.loadAlbumArt(uri: uri, size: size, cancellationSignalId: signal._id);
+
+  /// Tries to tell the system to recreate album art by [albumId].
+  ///
+  /// Sometimes `MediaStore` tells that there's an album art for some song, but the actual file
+  /// by some path doesn't exist. Supposedly, what happens is that Android detects reads on this
+  /// entry from something like `InputStream` in Java and regenerates album thumbs if they do not exist
+  /// (because this is just a caches that system tends to clear out if it thinks it should),
+  /// but (yet again, supposedly) this is not the case when Flutter accesses this file. System cannot
+  /// detect it and does not recreate the thumb, so we do this instead.
+  ///
+  /// See https://stackoverflow.com/questions/18606007/android-image-files-deleted-from-com-android-providers-media-albumthumbs-on-rebo
+  static Future<void> fixAlbumArt(int albumId) async => await SweyerPluginPlatform.instance.fixAlbumArt(albumId);
+
+  static Future<Iterable<T>> retrieveSongs<T extends MediaStoreSong>(
+    T Function(Map<String, dynamic> data) factory,
+  ) async =>
+      (await SweyerPluginPlatform.instance.retrieveSongs()).map(factory);
+
+  static Future<Map<int, T>> retrieveAlbums<T extends MediaStoreAlbum>(
+    T Function(Map<String, dynamic> data) factory,
```

Comment:
Let's factor this function signature into a `typedef _DataFactory<T> = T Function(Map<String, dynamic> data);`